framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -949,146 +953,159 @@ public Void visitMethodInvocation(MethodInvocationTree node, Void p) {
     }
 
     /**
-     * Checks all the preconditions of the method invocation or variable access {@code tree} with
-     * element {@code invokedElement}.
+     * Checks that all the given {@code preconditions} hold true immediately prior to
+     * the method invocation or variable access at {@code tree}.
+     *
+     * @param tree the Tree immediately prior to which the preconditions must hold true.
+     * @param preconditions the preconditions to be checked.
      */
-    protected void checkPreconditions(Tree tree,
-            Element invokedElement, boolean methodCall) {
-        checkPreconditions(tree, invokedElement, methodCall, null);
+    protected void checkPreconditions(Tree tree, Set<Pair<String, String>> preconditions) {
+        // This check is needed for the GUI effects and Units Checkers tests to pass.
+        // TODO: Remove this check and investigate the root cause.
+        if (preconditions.isEmpty()) {
+            return;
+        }
+
+        checkPreconditions(tree, atypeFactory.getNodeForTree(tree), preconditions);
     }
 
     /**
-     * Checks all the preconditions of the method invocation or variable access {@code tree} with
-     * element {@code invokedElement}.
+     * Checks that all the given {@code preconditions} hold true immediately prior to
+     * the method invocation or variable access at {@code node}.  Errors are reported
+     * with respect to {@code treeForErrorReporting}, which does not need to correspond to {@code node}.
+     *
+     * @param treeForErrorReporting the Tree used to report the error via checker.report.
+     * @param node the Node immediately prior to which the preconditions must hold true.
+     * @param preconditions the preconditions to be checked.
      */
-    protected void checkPreconditions(Tree tree,
-            Element invokedElement, boolean methodCall, Set<Pair<String, String>> additionalPreconditions) {
-        Set<Pair<String, String>> preconditions = invokedElement == null ?
-                new HashSet<Pair<String, String>>() :
-                contractsUtils.getPreconditions(invokedElement);
-
-        if (additionalPreconditions != null) {
-            preconditions.addAll(additionalPreconditions);
+    protected void checkPreconditions(Tree treeForErrorReporting,
+            Node node, Set<Pair<String, String>> preconditions) {
+        if (preconditions.isEmpty()) {
+            return;
         }
 
-        FlowExpressionContext flowExprContext = null;
+        FlowExpressionContext flowExprContext = getFlowExpressionContextFromNode(node);
+
+        if (flowExprContext == null) {
+            checker.report(Result.failure("flowexpr.parse.context.not.determined",
+                    node), treeForErrorReporting);
+            return;
+        }
 
         for (Pair<String, String> p : preconditions) {
             String expression = p.first;
             AnnotationMirror anno = AnnotationUtils.fromName(elements, p.second);
```

Comment:
Instead of creating the AnnotationMirror here, why isn't the second element of a pre-condition pair the AnnotationMirror?